File: core/src/main/java/cloud/vinh/rebirthsaga/game/commands/CommandResult.java
package cloud.vinh.rebirthsaga.game.commands;

/** Outcome of one resolved command. Expected rejections are values, not
 * exceptions; only invariant failures should halt a session. Immutable. */
public final class CommandResult {
    public enum Reason {
        /** The command changed authoritative state. */
        ACCEPTED,
        /** The move delta was not a single cardinal step. */
        NOT_CARDINAL,
        /** The destination cell is outside the current floor. */
        OUT_OF_BOUNDS,
        /** The destination cell is terrain the actor cannot enter. */
        BLOCKED
    }

    public static final CommandResult ACCEPTED = new CommandResult(Reason.ACCEPTED);
    public static final CommandResult NOT_CARDINAL = new CommandResult(Reason.NOT_CARDINAL);
    public static final CommandResult OUT_OF_BOUNDS = new CommandResult(Reason.OUT_OF_BOUNDS);
    public static final CommandResult BLOCKED = new CommandResult(Reason.BLOCKED);

    private final Reason reason;

    private CommandResult(Reason reason) {
        this.reason = reason;
    }

    public Reason reason() {
        return reason;
    }

    public boolean accepted() {
        return reason == Reason.ACCEPTED;
    }

    @Override
    public String toString() {
        return "CommandResult(" + reason + ')';
    }
}

File: core/src/main/kotlin/cloud/vinh/rebirthdungeon/game/commands/CommandResult.kt
package cloud.vinh.rebirthdungeon.game.commands

/** Outcome of one resolved command. Expected rejections are values, not
 * exceptions; only invariant failures should halt a session. Immutable. */
data class CommandResult private constructor(val reason: Reason) {
    enum class Reason {
        /** The command changed authoritative state. */
        ACCEPTED,
        NOT_PLAYER_TURN, STALE_SESSION, SAVE_REQUIRED, INVALID_PHASE, INVALID_DICE, INVALID_SKILL, INVALID_TARGET, EQUIPMENT_REQUIRED, COOLDOWN, INSUFFICIENT_HP, INSUFFICIENT_MP, INSUFFICIENT_SP, TERMINAL, ITEM_UNAVAILABLE
    }

    fun accepted(): Boolean = reason == Reason.ACCEPTED

    override fun toString(): String = "CommandResult($reason)"

    companion object {
        fun rejected(reason: Reason) = CommandResult(reason)
        val ACCEPTED = CommandResult(Reason.ACCEPTED)
    }
}
